### PR TITLE
feat: Support valueType and optionSet in /eventReports [TECH-843]

### PR DESCRIPTION
**Pull request by maikelarabori**
This change adds two transient attributes in the response of the `/eventReports` endpoint.
They are `valueType` and `optionSet`.

These new attributes are populated in the response phase and are based on the loaded EventReport. They are applicable only for `TrackedEntityAttributeDimension` (attributeDimensions) and `TrackedEntityDataElementDimension` (dataElementDimensions).

They are being required as part of the new Event Visualization app.

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObject.java [hunk 1]
```diff
@@ -133,6 +135,16 @@
      */
     String getDimensionDisplayName();
 
+    /**
+     * Returns the value type of the dimension.
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObject.java:139]: @maikelarabori could you add a note in the Javadoc that not all implementing classes have a ValueType field, and clarify the behavior in those cases (e.g. null can be returned).

**Review by gnespolino — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObject.java (hunk 1 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObject.java:139]: Clarification add @larshelge, thanks

Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/DimensionalObject.java:139]: Clarification added @larshelge, thanks

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 2]
```diff
@@ -700,47 +637,64 @@ else if ( STATIC_DIMS.contains( dimension ) )
...
+        final Map<String, TrackedEntityAttributeDimension> attributes = Maps.uniqueIndex( attributeDimensions,
+            TrackedEntityAttributeDimension::getUid );
 
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_DATA_ELEMENT, null,
-                    tedd.getDisplayName(), tedd.getLegendSet(), tedd.getProgramStage(), tedd.getFilter() );
-            }
+        if ( attributes.containsKey( dimension ) )
+        {
+            final TrackedEntityAttributeDimension tead = attributes.get( dimension );
+            final ValueType valueType = tead.getAttribute() != null ? tead.getAttribute().getValueType() : null;
```
Comment (d-bernat) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:661]: tead still can be null, can you please add the test to avoid null ptr exception?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 3]
```diff
@@ -700,47 +637,64 @@ else if ( STATIC_DIMS.contains( dimension ) )
...
-                TrackedEntityProgramIndicatorDimension teid = programIndicators.get( dimension );
+        final Map<String, TrackedEntityDataElementDimension> dataElements = Maps.uniqueIndex( dataElementDimensions,
+            TrackedEntityDataElementDimension::getUid );
 
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_INDICATOR, null,
-                    teid.getDisplayName(), teid.getLegendSet(), null, teid.getFilter() );
-            }
+        if ( dataElements.containsKey( dimension ) )
+        {
+            final TrackedEntityDataElementDimension tedd = dataElements.get( dimension );
```
Comment (d-bernat) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:675]: tedd still can be null, can you please add the test to avoid null ptr exception?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 4]
```diff
@@ -700,47 +637,64 @@ else if ( STATIC_DIMS.contains( dimension ) )
...
 
-        throw new IllegalArgumentException( "Not a valid dimension: " + dimension );
+        // Tracked entity program indicator
+
+        final Map<String, TrackedEntityProgramIndicatorDimension> programIndicators = Maps
+            .uniqueIndex( programIndicatorDimensions, TrackedEntityProgramIndicatorDimension::getUid );
+
+        if ( programIndicators.containsKey( dimension ) )
+        {
+            final TrackedEntityProgramIndicatorDimension teid = programIndicators.get( dimension );
```
Comment (d-bernat) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:691]: teid still can be null, can you please add the test to avoid null ptr exception?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java (hunk 2 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java]: This was legacy code, but doesn't hurt adding this check just to play on the safe side :+1: 
Change applied.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java (hunk 3 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:681]: Same here.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java (hunk 4 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:705]: And also here

**Comment by maikelarabori** [comment]
Hi @d-bernat, changes applied. Please take another look once you have time. Many thanks! :+1:

**Review by d-bernat — APPROVED** [review]
